libraries/Microsoft.Bot.Connector/OAuthApiClient.cs

```diff
@@ -0,0 +1,350 @@
+            }
+            else
+            {
+                tokenUrl = tokenUrl.Replace("{magicCodeParam}", String.Empty);
+            }
+
+            // Create HTTP transport objects
+            var httpRequest = new HttpRequestMessage();
+            HttpResponseMessage httpResponse = null;
+            httpRequest.Method = new HttpMethod("GET");
+            httpRequest.RequestUri = new Uri(tokenUrl);
+
+            // add botframework api service url to the list of trusted service url's for these app credentials.
+            MicrosoftAppCredentials.TrustServiceUrl(tokenUrl);
+
+            // Set Credentials
+            if (client.Credentials != null)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                await client.Credentials.ProcessHttpRequestAsync(httpRequest, cancellationToken).ConfigureAwait(false);
+            }
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (shouldTrace)
+            {
+                ServiceClientTracing.SendRequest(invocationId, httpRequest);
+            }
+            httpResponse = await client.HttpClient.SendAsync(httpRequest, cancellationToken).ConfigureAwait(false);
+            if (shouldTrace)
+            {
+                ServiceClientTracing.ReceiveResponse(invocationId, httpResponse);
+            }
+            HttpStatusCode statusCode = httpResponse.StatusCode;
+            cancellationToken.ThrowIfCancellationRequested();
+            if (statusCode == HttpStatusCode.OK)
+            {
+                string responseContent = await httpResponse.Content.ReadAsStringAsync().ConfigureAwait(false);
+                try
+                {
+                    var tokenResponse = Rest.Serialization.SafeJsonConvert.DeserializeObject<TokenResponse>(responseContent);
+                    return tokenResponse;
+                }
+                catch (JsonException)
+                {
+                    httpRequest.Dispose();
+                    if (httpResponse != null)
+                    {
+                        httpResponse.Dispose();
+                    }
+                    return null;
+                }
+            }
+            else if (statusCode == HttpStatusCode.NotFound)
+            {
+                return null;
+            }
+            else
+            {
+                return null;
+            }
+        }
+        
+        /// <summary>
+        /// Signs Out the User for the given ConnectionName.
+        /// </summary>
+        /// <param name="userId"></param>
+        /// <param name="connectionName"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<bool> SignOutUserAsync(string userId, string connectionName, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (string.IsNullOrEmpty(userId))
```

Comment:
You've got these all throughout the code. Consider adding:
private static void AssertValidUserId(string userId)
{
  if (string.isNullOrWhitespace(userId)) throw new ArgumentNullExcpetion(nameof(userId))
}

Reply:
it's in just a couple of places and more descriptive, so i dont think it's worth encapsulating in anohter method.